x-pack/plugin/migrate/src/main/java/org/elasticsearch/xpack/migrate/action/ReindexDataStreamIndexTransportAction.java

```diff
@@ -0,0 +1,240 @@
+import org.elasticsearch.index.reindex.ReindexAction;
+import org.elasticsearch.index.reindex.ReindexRequest;
+import org.elasticsearch.injection.guice.Inject;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class ReindexDataStreamIndexTransportAction extends HandledTransportAction<
+    ReindexDataStreamIndexAction.Request,
+    ReindexDataStreamIndexAction.Response> {
+
+    private static final Logger logger = LogManager.getLogger(ReindexDataStreamIndexTransportAction.class);
+
+    private static final Set<String> SETTINGS_TO_ADD_BACK = Set.of(IndexMetadata.SETTING_BLOCKS_WRITE, IndexMetadata.SETTING_READ_ONLY);
+
+    private static final IndicesOptions IGNORE_MISSING_OPTIONS = IndicesOptions.fromOptions(true, true, false, false);
+    private final ClusterService clusterService;
+    private final Client client;
+    private final IndexScopedSettings indexScopedSettings;
+
+    @Inject
+    public ReindexDataStreamIndexTransportAction(
+        TransportService transportService,
+        ClusterService clusterService,
+        ActionFilters actionFilters,
+        Client client,
+        IndexScopedSettings indexScopedSettings
+    ) {
+        super(
+            ReindexDataStreamIndexAction.NAME,
+            false,
+            transportService,
+            actionFilters,
+            ReindexDataStreamIndexAction.Request::new,
+            transportService.getThreadPool().executor(ThreadPool.Names.GENERIC)
+        );
+        this.clusterService = clusterService;
+        this.client = client;
+        this.indexScopedSettings = indexScopedSettings;
+    }
+
+    @Override
+    protected void doExecute(
+        Task task,
+        ReindexDataStreamIndexAction.Request request,
+        ActionListener<ReindexDataStreamIndexAction.Response> listener
+    ) {
+        var sourceIndexName = request.getSourceIndex();
+        var destIndexName = generateDestIndexName(sourceIndexName);
+        IndexMetadata sourceIndex = clusterService.state().getMetadata().index(sourceIndexName);
+        Settings settingsBefore = sourceIndex.getSettings();
+
+        if (sourceIndex.getCreationVersion().isLegacyIndexVersion() == false) {
+            logger.warn(
+                "Reindexing datastream index with non-legacy version, index [{}], version [{}]",
+                sourceIndexName,
+                sourceIndex.getCreationVersion()
+            );
+        }
+
+        AtomicReference<BulkByScrollResponse> reindexResponse = new AtomicReference<>();
+        SubscribableListener.<AcknowledgedResponse>newForked(l -> setReadOnly(sourceIndexName, l))
+            .<AcknowledgedResponse>andThen(l -> deleteDestIfExists(request.getDeleteDestIfExists(), destIndexName, l))
```

Comment:
Should we actually delete the dest index if it exists? It seems we could just keep adding docs to the existing one. Though perhaps deleting is safer as it might clean up some issue that occurred in a previous run :man_shrugging:

Reply:
Yeah I would think deleting it might be safer because that way we'd pick up any mapping changes a user has made. Although then do we run the risk of some rogue reindex process accidentally creating the dest index just by writing to it? So it would have no mappings or settings?

Reply:
Hmm, good point about the possibility of the dest index getting created without mappings. Maybe we should instead fail if the index exists when we try to create it. This would require a rerun to fix the issue (eg run delete then create again). But I think this would be very rare, as it would require that another action create the index between this action deleting any existing dest index and then immediately creating the index. What do you think?